### ISSUE TITLE
importing joplin notes doesn't create sub-folders

### ISSUE THREAD

**Issue by pcause** [pbek/QOwnNotes#2361]
A comment for 21.8.2 version seems to imply that Joplin import will preserve folder structure if the not folder has sub-folders enabled.  I imported a joplin export into my default (only) note folder.  The joplin export was one where there were sub-folders.  I have allow sub-folder enabled.  No new sub-folders were created for the Joplin sub-folders  All notes were imported.  looking at the exported joplin files, there are parent links in the end matter to notes that are the folders.

Regular notes seem to be type_: 1 and folders type_: 2 and the folders I see in Joplin have md files with this end matter.
.

**Comment by pbek** [comment]
Did you enable sub folders for your note folder? (you never fill out the issue template, it's hard to help you 😅)

**Comment by pcause** [comment]
sorry I don't and will try next.  yes, I had sub-folders enabled.  This is on Windows 11.  Haven't tried on Linux yet.

**Comment by pbek** [comment]
Can you please post the output from the debug settings that you can copy when you get into the settings dialog in QOwnNotes and head over to the *Debug* section of it.

Do you have an example Joplin raw export folder where you had the issue?

**Comment by pcause** [comment]
Here is the folder settings:

![image](https://user-images.githubusercontent.com/1071479/147967716-1ee8bbd4-2d53-4c96-a975-5866d0e50512.png)

her is the debug setting
[folderimportsettings.zip](https://github.com/pbek/QOwnNotes/files/7803530/folderimportsettings.zip)

here is the simple test i made and exported from joplin
[simpledirs.zip](https://github.com/pbek/QOwnNotes/files/7803532/simpledirs.zip)

here is the debug panel

[13:44:24] [debug] handleNoteTextFileName - 'name' was changed: "The Best Restaurants In The South End - Boston - The Infatuation - Copy"
[13:44:24] [debug] storing note file: "The Best Restaurants In The South End - Boston - The Infatuation - Copy.md"
[13:44:24] [debug] storing note file: "The Best Restaurants In The South End - Boston - The Infatuation - Copy.md"
[13:44:24] [debug] storing note file: "The Best Restaurants In The South End - Boston - The Infatuation - Copy.md"
[13:44:24] [debug] handleNoteTextFileName - 'name' was changed: "No More Chuuka | Dining | Metropolis Japan"
[13:44:24] [debug] handleNoteTextFileName - cannot write to file "No More Chuuka | Dining | Metropolis Japan.md" - we will try another filename
[13:44:24] [debug] storing note file: "No More Chuuka Dining Metropolis Japan.md"
[13:44:24] [debug] importImage - 'imageName': "covid-19"
[13:44:24] [debug] importImage - 'imageId': "a2246637a29e42d6a314b4e35cef4457"
[13:44:24] [debug] importImage - 'mediaFile': QFile(0x8dfbd20)
[13:44:24] [debug] findAvailableFileName - 'override fileName': "a2246637a29e42d6a314b4e35cef4457-1.jpg"
[13:44:24] [debug] findAvailableFileName - 'override fileName': "a2246637a29e42d6a314b4e35cef4457-2.jpg"
[13:44:24] [debug] findAvailableFileName - 'override fileName': "a2246637a29e42d6a314b4e35cef4457-3.jpg"
[13:44:24] [debug] findAvailableFileName - 'override fileName': "a2246637a29e42d6a314b4e35cef4457-4.jpg"
[13:44:24] [debug] findAvailableFileName - 'override fileName': "a2246637a29e42d6a314b4e35cef4457-5.jpg"
[13:44:24] [debug] importImage - 'mediaMarkdown': "![covid-19](media/a2246637a29e42d6a314b4e35cef4457-5.jpg)"
[13:44:24] [debug] importImage - 'imageName': "kaenzan-tokyo-japan-metropolis"
[13:44:24] [debug] importImage - 'imageId': "2403aa9c027b422aa7306179db9f8674"
[13:44:24] [debug] importImage - 'mediaFile': QFile(0x8dfbfa0)
[13:44:24] [debug] findAvailableFileName - 'override fileName': "2403aa9c027b422aa7306179db9f8674-1.jpg"
[13:44:24] [debug] findAvailableFileName - 'override fileName': "2403aa9c027b422aa7306179db9f8674-2.jpg"
[13:44:24] [debug] findAvailableFileName - 'override fileName': "2403aa9c027b422aa7306179db9f8674-3.jpg"
[13:44:24] [debug] findAvailableFileName - 'override fileName': "2403aa9c027b422aa7306179db9f8674-4.jpg"
[13:44:24] [debug] findAvailableFileName - 'override fileName': "2403aa9c027b422aa7306179db9f8674-5.jpg"
[13:44:24] [debug] importImage - 'mediaMarkdown': "![kaenzan-tokyo-japan-metropolis](media/2403aa9c027b422aa7306179db9f8674-5.jpg)"
[13:44:24] [debug] importImage - 'imageName': "kaiteirou-tokyo-japan-metropolis"
[13:44:24] [debug] importImage - 'imageId': "cb0e383a62c24f3f9bd5197c3dbe14f3"
[13:44:24] [debug] importImage - 'mediaFile': QFile(0x8dfbd00)
[13:44:24] [debug] findAvailableFileName - 'override fileName': "cb0e383a62c24f3f9bd5197c3dbe14f3-1.jpeg"
[13:44:24] [debug] findAvailableFileName - 'override fileName': "cb0e383a62c24f3f9bd5197c3dbe14f3-2.jpeg"
[13:44:24] [debug] findAvailableFileName - 'override fileName': "cb0e383a62c24f3f9bd5197c3dbe14f3-3.jpeg"
[13:44:24] [debug] findAvailableFileName - 'override fileName': "cb0e383a62c24f3f9bd5197c3dbe14f3-4.jpeg"
[13:44:24] [debug] findAvailableFileName - 'override fileName': "cb0e383a62c24f3f9bd5197c3dbe14f3-5.jpeg"
[13:44:24] [debug] importImage - 'mediaMarkdown': "![kaiteirou-tokyo-japan-metropolis](media/cb0e383a62c24f3f9bd5197c3dbe14f3-5.jpeg)"
[13:44:24] [debug] storing note file: "No More Chuuka Dining Metropolis Japan.md"
[13:44:24] [debug] storing note file: "No More Chuuka Dining Metropolis Japan.md"
[13:44:24] [debug] handleNoteTextFileName - 'name' was changed: "Sizzling Korean Barbecue Restaurants to Try in NYC"
[13:44:24] [debug] storing note file: "Sizzling Korean Barbecue Restaurants to Try in NYC.md"
[13:44:24] [debug] storing note file: "Sizzling Korean Barbecue Restaurants to Try in NYC.md"
[13:44:24] [debug] storing note file: "Sizzling Korean Barbecue Restaurants to Try in NYC.md"
[13:44:28] [debug] buildNotesIndex - 'noteSubFolderId': 0
[13:44:28] [debug] setupNoteFolderTables - 'database version': 15
[13:44:28] [debug] cloudConnection: CloudConnection: <id>1 <name>"" <serverUrl>"" <username>"" <accountId>"" <priority>0
[13:44:28] [debug] loadNoteDirectoryList
[13:44:28] [debug] reloadTagTree
[13:44:28] [debug] reloadTagTree - 'noteSubFolderIds': QVector()
[13:44:28] [debug] filterNotesByTag - 'tags': QVector()
[13:44:28] [debug] filterNotesByTag - 'noteIdList': QVector(1)
[13:44:28] [debug] reloadNoteSubFolderTree
[13:44:28] [debug] filterNotesByTag - 'tags': QVector()
[13:44:28] [debug] filterNotesByTag - 'noteIdList': QVector(1)
[13:44:28] [debug] filterNotesByTag - 'tags': QVector()
[13:44:28] [debug] filterNotesByTag - 'noteIdList': QVector(1)
[13:44:28] [debug] reloadTagTree
[13:44:28] [debug] reloadTagTree - 'noteSubFolderIds': QVector()
[13:44:34] [debug] filterNotesByNoteSubFolders - 'noteSubFolderIds': QVector(0)
[13:44:34] [debug] filterNotesByTag - 'tags': QVector()
[13:44:34] [debug] filterNotesByTag - 'noteIdList': QVector(1)
[13:44:34] [debug] reloadTagTree
[13:44:34] [debug] reloadTagTree - 'noteSubFolderIds': QVector()
[13:44:40] [debug] cloudConnection: CloudConnection: <id>1 <name>"" <serverUrl>"" <username>"" <accountId>"" <priority>0
[13:46:50] [debug] initShortcuts - 'menuBarIsVisible': true
[13:46:50] [debug] cloudConnection: CloudConnection: <id>1 <name>"" <serverUrl>"" <username>"" <accountId>"" <priority>0
[13:46:50] [debug] buildNotesIndex - 'noteSubFolderId': 0
[13:46:50] [debug] setupNoteFolderTables - 'database version': 15
[13:46:50] [debug] cloudConnection: CloudConnection: <id>1 <name>"" <serverUrl>"" <username>"" <accountId>"" <priority>0
[13:46:50] [debug] loadNoteDirectoryList
[13:46:50] [debug] reloadTagTree
[13:46:50] [debug] reloadTagTree - 'noteSubFolderIds': QVector(0)
[13:46:50] [debug] filterNotesByNoteSubFolders - 'noteSubFolderIds': QVector(0)
[13:46:50] [debug] filterNotesByTag - 'tags': QVector()
[13:46:50] [debug] filterNotesByTag - 'noteIdList': QVector(1)
[13:46:50] [debug] reloadNoteSubFolderTree
[13:46:50] [debug] filterNotesByNoteSubFolders - 'noteSubFolderIds': QVector(0)
[13:46:50] [debug] filterNotesByTag - 'tags': QVector()
[13:46:50] [debug] filterNotesByTag - 'noteIdList': QVector(1)
[13:46:50] [debug] reloadNoteSubFolderTree
[13:46:50] [debug] loadNoteDirectoryList
[13:46:50] [debug] reloadTagTree
[13:46:50] [debug] reloadTagTree - 'noteSubFolderIds': QVector(0)
[13:46:50] [debug] filterNotesByNoteSubFolders - 'noteSubFolderIds': QVector(0)
[13:46:50] [debug] filterNotesByTag - 'tags': QVector()
[13:46:50] [debug] filterNotesByTag - 'noteIdList': QVector(1)
[13:46:50] [debug] reloadNoteSubFolderTree
[13:46:50] [debug] filterNotesByNoteSubFolders - 'noteSubFolderIds': QVector(0)
[13:46:50] [debug] filterNotesByTag - 'tags': QVector()
[13:46:50] [debug] filterNotesByTag - 'noteIdList': QVector(1)

here is the screen (I did a reload folder to be sure)

![image](https://user-images.githubusercontent.com/1071479/147968045-37007cc3-c9d7-4783-87b0-d81bc9767631.png)

Hope this helps.

**Comment by pbek** [comment]
> here is the debug panel

that's the log panel (but thanks, that was also part of the issue template)

Can you please post the output from the debug settings that you can copy when you get into the settings dialog in QOwnNotes and head over to the *Debug* section of it.

**Comment by pcause** [comment]
> > here is the debug panel
> 
> that's the log panel (but thanks, that was also part of the issue template)
> 
> Can you please post the output from the debug settings that you can copy when you get into the settings dialog in QOwnNotes and head over to the _Debug_ section of it.

the debug panel iszipped above because we can't attached a text file folderimportsettings.zip

**Comment by pbek** [comment]
> a0bf2fd489e148b18399482d3f783f8e.md:parent_id: 1d77eb97c1b244489132a7a62009c117

Are you sure every folder is correctly exported from joplin? Already the first folder `1d77eb97c1b244489132a7a62009c117` that is supposed to be created is not present in your raw export folder as `1d77eb97c1b244489132a7a62009c117.md`. That might cause issues...

**Comment by pcause** [comment]
I see md items for the 3 folder and the top folder that contains them.The one you mention above it the top level I exported from - called qown.  Its parent is the root of the notebook.   I think if you find a directory without a parent in an export it was the top level exported and should also be a sub-folder.  What you should have is:

```
qown
- folder1
- folder2
- -folder3
```

and so in default I should see qown as a subfolder and folder1, folder2 and folder3 as sibfolders in qown

**Comment by pbek** [comment]
I'm not sure if the case when referenced items are missing can be handled, if you for example take the export from https://github.com/pbek/QOwnNotes/issues/2189#issuecomment-882083064, which also contains folders and import it you can see all the folders are created. Why should there a parent_id referenced that don't exist in the export?

**Comment by pbek** [comment]
In the example from https://github.com/pbek/QOwnNotes/issues/2189#issuecomment-882083064 a folder markdown file may look like this:

```markdown
Welcome! (Desktop)

id: 592a89c5cd064b65974ac5cb9e1abc72
created_time: 2021-07-18T06:51:38.047Z
updated_time: 2021-07-18T06:51:38.047Z
user_created_time: 2021-07-18T06:51:38.047Z
user_updated_time: 2021-07-18T06:51:38.047Z
encryption_cipher_text: 
encryption_applied: 0
parent_id: 
is_shared: 0
share_id: 
type_: 2
```

No `parent_id`, because it's a root folder.

**Comment by pcause** [comment]
in joplin you can start an export at any level of folder tree.  so if you get an exported directory and it has a folder (type 2) that has a parent that doesn't exist, it is the root of the export.   so make that subfolder a subfolder and then all the other will work

**Comment by pbek** [comment]
## 22.1.3
- Joplin sub-folders can now also be imported if the parent folder is missing

**Comment by pbek** [comment]
There now is a new release, could you please test it and report if it works for you?

**Comment by pcause** [comment]
fixed in 22.1.3

**Comment by pbek** [comment]
Great, thank you for testing!
